lib/core/src/Cardano/Wallet/Primitive/Model.hs

```diff
@@ -411,58 +444,43 @@ utxoFromUnvalidatedTx :: Tx -> UTxO
 utxoFromUnvalidatedTx Tx {txId, outputs} =
     UTxO $ Map.fromList $ zip (TxIn txId <$> [0..]) outputs
 
-isOurAddress
-    :: IsOurs s Address
-    => Address
-    -> State s Bool
-isOurAddress = fmap isJust . state . isOurs
-
-isOurWithdrawal
-    :: IsOurs s RewardAccount
-    => (RewardAccount, Coin)
-    -> State s Bool
-isOurWithdrawal = fmap isJust . ourWithdrawal
-
-ourDelegation
-    :: IsOurs s RewardAccount
-    => DelegationCertificate
-    -> State s (Maybe DelegationCertificate)
-ourDelegation cert =
-    state (isOurs $ dlgCertAccount cert) <&> \case
-        Nothing -> Nothing
-        Just{}  -> Just cert
-
-ourWithdrawal
-    :: IsOurs s RewardAccount
-    => (RewardAccount, Coin)
-    -> State s (Maybe (RewardAccount, Coin))
-ourWithdrawal (acct, amt) =
-    state (isOurs acct) <&> \case
-        Nothing -> Nothing
-        Just{}  -> Just (acct, amt)
-
-ourWithdrawalSumFromTx
-    :: IsOurs s RewardAccount
-    => Tx
-    -> State s Coin
-ourWithdrawalSumFromTx tx
-    -- If a transaction has failed script validation, then the ledger rules
-    -- require that applying the transaction shall have no effect other than
-    -- to fully spend the collateral inputs included within that transaction.
-    --
-    -- Therefore, any reward withdrawals included in such a transaction should
-    -- also have no effect.
-    --
-    | failedScriptValidation tx =
-        pure (Coin 0)
-    | otherwise =
-        F.foldMap snd <$>
-        mapMaybeM ourWithdrawal (Map.toList $ withdrawals tx)
-
--- | Indicates whether a given transaction is relevant to the wallet.
---
+{-------------------------------------------------------------------------------
+    Address ownership and discovery
+-------------------------------------------------------------------------------}
+-- | Perform address discovery by going through all transactions
+-- and delegation certificates in the block.
+discoverAddresses
+    :: (IsOurs s Address, IsOurs s RewardAccount)
+    => Block -> s -> s
+discoverAddresses block s0 = s2
+  where
+    -- NOTE: Order in which we perform discovery is important.
+    s1 = L.foldl' discoverCert s0 (block ^. #delegations)
+    s2 = L.foldl' discoverTx   s1 (block ^. #transactions)
+
+    discoverCert s cert = updateOurs s (dlgCertAccount cert)
+
+    -- NOTE: Only outputs and withdrawals can introduce new addresses.
+    -- Inputs and collateral are forced to use existing addresses.
+    discoverTx s tx = discoverWithdrawals (discoverOutputs s tx) tx
+    discoverOutputs s tx =
+        L.foldl' (\s_ out -> updateOurs s_ (out ^. #address)) s (tx ^. #outputs)
+    discoverWithdrawals s tx =
+        L.foldl' updateOurs s $ Map.keys (tx ^. #withdrawals)
+
+-- | Helper based on 'isOurs'.
+updateOurs :: IsOurs s entity => s -> entity -> s
```

Comment:
is updateOurs used elsewhere? If not then maybe we can put it  where of discoverAddresses

Reply:
I was thinking the same, but `updateOurs` is sort of the sibling of `ours` which needs to be at the top level.

Reply:
> `updateOurs` is sort of the sibling of `ours` which needs to be at the top level.

Yes. I actually intend to move these functions into the `IsOurs` class in some future PR.